### PR TITLE
Protobuf generation + ONNX ML extension

### PR DESCRIPTION
**Pull request by ezyang**
The basic design parameters:

- For now, we will check in the generated build products (we can
  stop doing this when we have a way of publishing the generated
  protobuf when it's ready.

- The generator is done in Python.

We can decide exactly how we want to turn the crank on the actual
transformation when we rebase the ONNX-ML changes on top.

CC @donbox, what is the most up-to-date version of ONNX with your changes for CoreML? If I have those changes, I can figure out what kind of text processing we need to generate both versions of the proto.

### CODE REVIEW

**Comment by gramalingam** [comment]
I have a basic question: what is the compatibility goal/requirement between the two protobuf definitions?

**Comment by ezyang** [comment]
@gramalingam They need to be binary compatible. We can reverse engineer this using the backward compatibility rules (https://developers.google.com/protocol-buffers/docs/proto) which say: (1) don't change any numeric tags, (2) make all fields optional. So, as long as the transformation from ONNX to ONNX-ML is a "backwards-compatible change", the two protobufs are compatible.

**Comment by donbox** [comment]
Sorry for the delayed reply - I've been sick and am just now starting to be lucid enough to do any technical work.

I just updated dbox/add_map_record_sequence (a.k.a. PR67) to clearly delinate what we'd trim or not depending on the target.  

I simply used #if/#endif inside of a proto // comment as a placeholder.

I'm less worried about the python generator (which as you say, is a no-op).

What we should get agreement on is:

1. Is onnx.proto.in a valid proto file, at least syntactically?   As someone who edits the file, I'd prefer that my text editor's language services don't break if we can achieve that.

2. For conditional inclusion of specific declarations (e.g., messages, fields, etc), I think we can use an embedded-in-comments ifdef syntax. Honestly, that will keep us honest wrt colliding field IDs.  We should decide whether we want #if or #iddef semantics. For our purposes, I'd strongly favor #ifdef semantics.

3. I don't have a good answer for how to deal with the optional keyword. If we are going to treat proto2-vs-proto3 as a first-class thing, I assume we could simply bulk inject it (or bulk remove it).  I'm loath to add a general #define.

**Comment by donbox** [comment]
I think we should really use the proto2/proto3 thing to vet this design.   @ezyang  are you up for combining that with this PR?   Solving for the PR67/ONNX-ML stuff is trivial compared to solving that problem.

**Comment by dzhulgakov** [comment]
One suggestion - add verification to Travis similarly to #136?

**Comment by ezyang** [comment]
OK, it is now wired up to do ONNX/ONNX-ML output, as well as proto2/proto3 output. I based this on top of PR #67. CI is not done yet and will be added.

**Comment by prasanthpul** [comment]
@ezyang are you adding those tests or you want someone else to contribute them?

File: onnx/gen_proto.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        # Set the syntax specifier
+        m = PROTO_SYNTAX_REGEX.match(line)
+        if m:
+            yield m.group(1) + 'syntax = "proto3";'
+            continue
+
+        # Remove optinoal keywords
+        m = OPTIONAL_REGEX.match(line)
+        if m:
+            yield m.group(1) + m.group(2)
```
Comment (prasanthpul) [inline onnx/gen_proto.py]: nit: when "optional" is stripped the trailing whitespace is left behind causing different indentation for proto3 vs proto2 versions.

**Comment by ezyang** [comment]
I spent some time trying to add them, but it seems to be irritating to do this in a robust way. I wanted the test to be invariant to proto changes; i.e., you could change your proto, and if you introduced a breaking feature, the test would automatically flag it; you wouldn't have to remember to update some test files or something like that. But this rules out having some pre-computed model files that you test for input in proto2 and proto3 (since, definitionally, they won't test new features.) I experimented with https://github.com/trailofbits/protofuzz for randomly generating protobuf, but their generation strategy is not robust enough (for one, it infinite loops on recursive structs.)

**Comment by ezyang** [comment]
@prasanthpul Nice catch! I pushed a commit that fixes it.

**Comment by prasanthpul** [comment]
@ezyang let's keep more automated tests for proto3 as a separate item

**Comment by ezyang** [comment]
As soon as CI is green I intend to merge this.

**Pull request merged**